src/main/java/bio/terra/service/dataset/flight/upgrade/predictableFileIds/ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep.java

```diff
@@ -0,0 +1,50 @@
+package bio.terra.service.dataset.flight.upgrade.predictableFileIds;
+
+import bio.terra.service.dataset.Dataset;
+import bio.terra.service.dataset.DatasetService;
+import bio.terra.service.filedata.google.firestore.FireStoreDao;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep implements Step {
+
+  private final UUID datasetId;
+  private final DatasetService datasetService;
+  private final FireStoreDao fileDao;
+
+  public ConvertToPredictableFileIdsUpdateFirestoreCollectionsStep(
+      UUID datasetId, DatasetService datasetService, FireStoreDao fileDao) {
+    this.datasetId = datasetId;
+    this.datasetService = datasetService;
+    this.fileDao = fileDao;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    Dataset dataset = datasetService.retrieve(datasetId);
+    Map<UUID, UUID> oldToNewMappings =
+        ConvertFileIdUtils.readFlightMappings(context.getWorkingMap());
+
+    fileDao.moveFileMetadata(dataset, oldToNewMappings);
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    Dataset dataset = datasetService.retrieve(datasetId);
+    // Reverse the id mapping to revert
+    Map<UUID, UUID> oldToNewMappings =
+        ConvertFileIdUtils.readFlightMappings(context.getWorkingMap()).entrySet().stream()
+            .collect(Collectors.toMap(Entry::getValue, Entry::getKey));
```

Comment:
Is it guaranteed that all values are unique, e.g., that you never have two different file IDs that both point at the same predictable ID? If there are collisions then this will only reverse one of the move operations.

Reply:
Yeah, the idea is that the ids are all 1:1

Reply:
So you couldn't have two files with different file IDs but with identical hashes? I thought that the original file ID mechanism was random UUID, and the new one is hashcode. So unless you're already mapping identically hashing files to the same random UUID, it seems like you could end up with different "unpredictable" file IDs that have the same predictable file ID.

Reply:
Ah so the file id mechanism isn't _just_ the hash.  It's also the size and the target path.  So you can (and often do) have the same file but with a different path.  We treat that as a different file